Add Version Check for CSR Verification

erbsland-dev opened the pull request:
Fixes #5738: This change introduces a check for the version number of a CSR document before its signature is verified. If the version number is not 1 (encoded as zero), the verification function fails with an `X509_R_UNSUPPORTED_VERSION` error.

To minimise impact, this check is only applied when verifying a certificate signing request using the `-verify` argument limiting it to the `X509_REQ_verify_ex` and `X509_REQ_verify` functions. This ensures that malformed certificate requests are rejected by a certification authority, enhancing security and preventing potential issues.

nhorman commented:
@erbsland-dev would you please sign an icla as documented [here](https://openssl.org/policies/cla.html) so that we can review, and potentially accept this PR?  I don't think we can circumvent the agreement, as it wouldn't be considered trivial.

nhorman reviewed (approved):
Looks good, thank you!

t8m commented on crypto/x509/x_all.c:
```diff
@@ -43,6 +43,11 @@ int X509_verify(X509 *a, EVP_PKEY *r)
                        const char *propq)
 {
+    if (X509_REQ_get_version(a) != X509_REQ_VERSION_1) {
+        ERR_raise_data(ERR_LIB_X509, ERR_R_UNSUPPORTED, "version != 1");
```
IMO `ERR_raise(ERR_LIB_X509, X509_R_UNSUPPORTED_VERSION);` would be more appropriate. Of course that requires running `make update` to add the new error reason.

erbsland-dev replied:
@t8m I fully agree. I avoided adding a new error reason to keep this pull request minimal and its impact easily observable.

However, since the library lacks version checks in most of the X509 code, I would recommend introducing this new error code along with stricter version checking for *all* X509 functions.

erbsland-dev commented:
@nhorman I also wrote a test for this, but I found no guidelines if the test shall go along with the original PR, or be committed as independent PR.

t8m commented:
> @nhorman I also wrote a test for this, but I found no guidelines if the test shall go along with the original PR, or be committed as independent PR.

It should be included in this PR.

erbsland-dev commented:
@t8m Based on your recommendation, I added the new error code X509_R_UNSUPPORTED_VERSION after assessing its impact. The addition of this error code for X509 has minimal impact but significantly improves the readability of the returned error message.

t8m commented on test/x509_req_test.c:
```diff
@@ -0,0 +1,81 @@
+    }
+    req = PEM_read_bio_X509_REQ(bio, NULL, 0, NULL);
+    if (req == NULL) {
+        ret = 1; // success, reading PEM with invalid CSR data is allowed to fail.
+        goto err;
+    }
+    if (!TEST_ptr(pkey = X509_REQ_get_pubkey(req))) {
+        goto err;
+    }
+    // Verification MUST fail at this point. ret != 1.
+    if (!TEST_int_ne(X509_REQ_verify(req, pkey), 1)) {
+        goto err;
+    }
+    ret = 1; // success
```
Please always use the C-style comments (not C++ style). Also please do not add `{}` to blocks of just one statement.

t8m commented on test/x509_req_test.c:
```diff
@@ -0,0 +1,81 @@
+    ADD_TEST(test_x509_req_detect_invalid_version);
+    return 1;
+}
+
```
Please drop this superfluous empty line. (one empty line between functions is sufficient)

t8m commented on test/x509_req_test.c:
```diff
@@ -0,0 +1,81 @@
+    BIO_free(bio);
+    return ret;
+}
+
```
Please drop this superfluous empty line. (one empty line between functions is sufficient)

t8m reviewed (changes requested):
Just style nits.

Also please drop the CLA: trivial annotations from the first commits.

erbsland-dev commented:
@t8m I apologize for the oversights; I’m not accustomed to writing C code anymore. I have also squashed the commits into two: one for the changes and one for the tests. Additionally, I removed the “CLA: trivial” note and adjusted the commit messages accordingly.

t8m commented:
@nhorman please rereview

erbsland-dev commented:
@t8m @nhorman If there are any smaller fixes or enhancements I can help with, please let me know. Since I don’t have a comprehensive overview of the project, it’s difficult for me to identify the most important issues. However, as a lifelong user of the library, I’m eager to contribute. If you can point me to specific tasks that need implementation, I’d be happy to work on them.

nhorman commented:
Review holds, approved

nhorman commented:
@erbsland-dev thank you for the offer.  To keep you informed, typically we mark issues that we feel should be addressed, but are not within our capacity to handle, with the "help wanted" tag.  You are more than welcome to scan that list at your discretion and create PR's for them at will.  If you like, leave a comment and I'll happily assign them to you.

Further, we have a "good first issue" tag that further refines out issues that have been in the past considered to be a good starting point for people just getting involved, if you're so inclined.

A word of warning, as you have likely noticed, I've been cleaning up our issue list lately, and the help wanted issues are not well groomed yet, so its a bit of a "wild west" situtation, and you may find many issues that are old, and not clearly defined.  I apologize for that, but do you're best, cleaning that up is on my list :)

openssl-machine commented:
This pull request is ready to merge

nhorman commented:
merged to master, thank you for your contribution!
